### DISCUSSION TITLE
Implementação de testes

### DISCUSSION THREAD

**Discussion by weescoelho** [diego3g/rsxp-2023#35] [General]
Olá pessoal como estão? 

Acredito que deveriamos tambem discutir a adoção de testes no projeto e qual ferramenta utilizar neste processo. Temos dois gigantes que é o nosso querido Jest e o Vitest. 

Decidindo a ferramenta tambem é importante já realizar o setup, assim como a confguração do  CI com o Github.

**Comment by CassianoJunior** [comment]
Como estamos usando o nest, acho que ele já vem com integração com o Jest, então acredito que seja com ele.

**Comment by Rodrigo001-dev** [comment]
Essa discussão estava falando sobre os testes, quem quiser dar uma olhada #14

**Reply by weescoelho** [reply]
Massa! Como foi fechada não tinha visto. Mas acredito que abordagem que eles estavam expondo seja a mais coerente. Manter os testes no  nível mais alto dado a velocidade que precisaremos ter no desenvolvimento.

**Comment by trolliama** [comment]
Acho que outra questão sobre os testes que talvez valha a pena discutir é: qual tipo de teste unitário deveríamos fazer.
 Existem dois tipos pelo que eu conheço, o London school e o Classic school.

Esse [site](https://freecontent.manning.com/what-is-a-unit-test-part-2-classical-vs-london-schools/) explica muito bem a diferença entre os dois. Mas em resumo é isto aqui:

**London school**: A escola de testes de unidade de Londres descreve que o isolamento do sistema em teste de seus colaboradores é fundamental. Isso significa que, se uma classe tiver dependência de outra classe ou várias classes, é necessário substituir todas essas dependências por test doubles

**Classical (detroit) school**: Nesse caso aqui ele toma uma medida diferente. Você pode em um teste unitário fazer uso de funcionalidades de mais de uma classe sem se preocupar em mockar ela, desde que essa classe não tenha alguma dependencia que possa acabar sendo compartilhada entre testes unitários diferente, como por exemplo alguma classe de repository. Como um repository faz uso de uma conexão com um banco de dados, esssa comunicação será compartilhada entre os testes e poderá causar erros por causa de modificações indevidas no BD. Nesse caso o repository deve ser mockado

Na minha opinião o estilo detroit é o mais confiável e mais fácil de manter pois não tem muitos mocks como no estilo london, mas digam o que vocês acham.

- Acredito que [este artigo](https://medium.com/@adrianbooth/test-driven-development-wars-detroit-vs-london-classicist-vs-mockist-9956c78ae95f) seja uma boa referência para o que estou falando.
- E, para quem se interessar, o livro "Unit Testing Principles, Practices, and Patterns" entra mais a fundo nessa questão e outras coisas
